src/realm/sync/noinst/protocol_codec.hpp

```diff
@@ -112,153 +113,124 @@ class ClientProtocol {
 
-            util::MemoryInputStream in;
-            in.unsetf(std::ios_base::skipws);
-            in.set_buffer(uncompressed_body.data(), uncompressed_body.data() + uncompressed_body_size);
-
             ReceivedChangesets received_changesets;
 
             // Loop through the body and find the changesets.
-            std::size_t position = 0;
-            while (position < uncompressed_body_size) {
-                version_type server_version;
-                version_type client_version;
-                timestamp_type origin_timestamp;
-                file_ident_type origin_file_ident;
-                std::size_t original_changeset_size, changeset_size;
-                char sp_1, sp_2, sp_3, sp_4, sp_5, sp_6;
-
-                in >> server_version >> sp_1 >> client_version >> sp_2 >> origin_timestamp >> sp_3 >>
-                    origin_file_ident >> sp_4 >> original_changeset_size >> sp_5 >> changeset_size >> sp_6;
-
-                bool good_syntax =
-                    in && sp_1 == ' ' && sp_2 == ' ' && sp_3 == ' ' && sp_4 == ' ' && sp_5 == ' ' && sp_6 == ' ';
-
-                if (!good_syntax) {
-                    logger.error("Bad changeset header syntax");
-                    connection.handle_protocol_error(Error::bad_changeset_header_syntax);
-                    return;
-                }
-
-                // Update position to the end of the change set
-                position = std::size_t(in.tellg()) + changeset_size;
-
-                if (position > uncompressed_body_size) {
-                    logger.error("Bad changeset size %1 > %2", position, uncompressed_body_size);
-                    connection.handle_protocol_error(Error::bad_changeset_size);
-                    return;
+            while (!body_data.empty()) {
+                realm::sync::Transformer::RemoteChangeset cur_changeset;
+                std::size_t changeset_size;
+                auto body_data_sw = parse_header_line(body_data, ' ', cur_changeset.remote_version,
+                                                      cur_changeset.last_integrated_local_version,
+                                                      cur_changeset.origin_timestamp, cur_changeset.origin_file_ident,
+                                                      cur_changeset.original_changeset_size, changeset_size);
+                if (!body_data_sw.is_ok()) {
+                    return report_error(Error::bad_changeset_header_syntax, "Bad changeset header syntax: %1",
+                                        body_data_sw.get_status());
                 }
+                body_data = body_data_sw.get_value();
 
-                if (server_version == 0) {
-                    // The received changeset can never have version 0.
-                    logger.error("Bad server version=0");
-                    connection.handle_protocol_error(Error::bad_server_version);
-                    return;
```

Comment:
This check seems to have gone missing, was that intentional?

Reply:
Oops. This check did go away in the FLX sync demo code that this PR is based off of, but I did not intend to remove it now. I'll put it back in.